Notes stored in separate table

dstandish opened the pull request.

dstandish committed:
wip

dstandish committed:
try revert rename

dstandish committed:
simplify

dstandish committed:
working, minimally

dstandish committed:
more reverting of notes -> note rename

dstandish committed:
more reverting of notes -> note rename

dstandish committed:
more reverting of notes -> note rename

dstandish committed:
remove scratch code

dstandish committed:
remove test speedup

dstandish committed:
restore admin view

dstandish committed:
add migration

dstandish committed:
add migration

dstandish committed:
tod

dstandish committed:
fix migration

dstandish committed:
Add DagRunNote

dstandish committed:
add migration file

dstandish committed:
disamble notes in search

dstandish committed:
fix dagrun tests

uranusjr commented on airflow/models/taskinstance.py:
```diff
@@ -2676,6 +2686,52 @@ def from_dict(cls, obj_dict: dict) -> SimpleTaskInstance:
         return cls(**obj_dict, start_date=start_date, end_date=end_date, key=ti_key)
 
 
+class TaskNote(Base):
```
Should this not be TaskInstanceNote?

dstandish replied:
yeah, should

dstandish committed:
fix some tests and tighten up relationships, i think

dstandish committed:
remove notes from create_dagrun method

dstandish committed:
more cleanup

dstandish committed:
fix collation

dstandish committed:
fix db cleanup test

dstandish committed:
more test fixup

dstandish committed:
more test fixup

dstandish committed:
rename to tinote

dstandish committed:
rename fixup

dstandish commented on airflow/www/utils.py:
```diff
@@ -717,10 +717,10 @@ def clean_column_names():
             if not isinstance(desc, AssociationProxy):
                 continue
-            proxy_instance = getattr(self.obj, desc.value_attr)
+            proxy_instance = getattr(self.obj, obj_attr)
```
previously the code assumed that the attr name where the proxy is stored on the object (`obj_attr`) is the same as the name of the field on the remote object (`desc.value_attr`).  but it doesn't have to be the case.  and here we store the reference to `task_instance_note.content` under the attr `notes`

ashb commented on airflow/models/dagrun.py:
```diff
@@ -66,6 +68,7 @@
 from airflow.utils.sqlalchemy import UtcDateTime, nulls_first, skip_locked, tuple_in_condition, with_row_locks
 from airflow.utils.state import DagRunState, State, TaskInstanceState
 from airflow.utils.types import NOTSET, ArgNotSet, DagRunType
+from airflow.www.fab_security.sqla.models import User
```
Hmmmm, not thrilled about having to import this here in our models (just for import speed reasons)

Let me have a play and see what I can do.

jedcunningham replied:
This is no longer being imported 👍

ashb commented on airflow/www/views.py:
```diff
@@ -263,19 +263,20 @@ def dag_to_grid(dag, dag_runs, session):
             TaskInstance.state,
-            TaskInstance.notes,
-            sqla.func.count(sqla.func.coalesce(TaskInstance.state, sqla.literal("no_status"))).label(
+            func.min(TaskInstanceNote.content).label("notes"),
```
Using `min` on a string field seems.... unexpectec.

dstandish replied:
so...
yes it's always weird when you see this but there's a reason
some dbs have `any_value` or `first`. that would make more sense here. but sqla no have this. `min` is proxy for "just give me something"
the alternative is what was there before, which is adding to the group by.  but that doesn't make much sense either (i asked brent, he said we don't need to be creating new groups based on the presence of notes).

ashb commented on tests/api_connexion/endpoints/test_dag_run_endpoint.py:
```diff
@@ -1421,7 +1421,7 @@ def test_should_respond_200(self, dag_maker, session):
             "execution_date": dr.execution_date.isoformat(),
             "external_trigger": False,
             "logical_date": dr.logical_date.isoformat(),
-            "start_date": dr.logical_date.isoformat(),
```
This feels like it shouldn't have changed.

ashb replied:
Okay this is right due to this code, and it shouldn't have haver been passing before! (it was missing a flush or a proper refresh before.)

https://github.com/apache/airflow/blob/fb7c6afc8cb7f93909bd2e654ea185eb6abcc1ea/airflow/models/taskinstance.py#L276-L278

dstandish replied:
yes it's a consequence of me reading the DR from the db now.  i think before it called `refresh_from_db` but that updates _very_ selectively.

dstandish replied:
before
https://github.com/apache/airflow/blob/fdb77141783b8dc4039353e1be165a7e0a42fc26/airflow/api_connexion/endpoints/dag_run_endpoint.py#L415

now
https://github.com/apache/airflow/blob/68696c2f7edf1e9727b4b4c348245c138e05038f/airflow/api_connexion/endpoints/dag_run_endpoint.py#L416

refresh from db code:
https://github.com/apache/airflow/blob/fb7c6afc8cb7f93909bd2e654ea185eb6abcc1ea/airflow/models/dagrun.py#L247-L255

dstandish replied:
but i didn't want to load the notes in all usages of `refresh_from_db`.  so better to not use the helper.

ashb commented on tests/api_connexion/endpoints/test_task_instance_endpoint.py:
```diff
@@ -380,18 +380,13 @@ def test_should_respond_200_task_instance_with_sla_and_rendered(self, session):
         tis = self.create_task_instances(session)
-        session.query()
-        ti = tis[0]
-        ti.map_index = 1
```
You removed the mapped Ti test. We probably shouldn't.

dstandish replied:
no i simplified

dstandish replied:
the issue that forced the simplification...
what the code was doing is this...

> we need a few mapped ti, so let's take existing ti, notes and all, and change the map index

but now that notes is diff table, you can't just change the map index of ti because that won't also change the index of the keyed obj, namely notes.

simpler to just create _new_ ti.

ashb committed:
Don't import FAB user models just to define FK rel

We don't (currently) define any relationships it's just for making the
FK match the migration, so for now we can have the FK col defined as a
string.

When we eventually add a relationship to the get the creator of the
note, we should move the FAB User model into airflow.models and change
Security manager code to import from there instead.

ashb committed:
Avoid touching test file unnecessarily

dstandish committed:
fix import

jedcunningham commented on airflow/models/__init__.py:
```diff
@@ -69,6 +69,7 @@ def import_all_models():
     import airflow.models.dataset
     import airflow.models.serialized_dag
     import airflow.models.tasklog
+    import airflow.www.fab_security.sqla.models
```
(note to self: I don't think this is needed, check it)

jedcunningham replied:
Yep, already done here:
https://github.com/apache/airflow/blob/fb7c6afc8cb7f93909bd2e654ea185eb6abcc1ea/airflow/utils/db.py#L667

dstandish replied:
It fails without that pretty sure

jedcunningham replied:
What's "it"?

jedcunningham replied:
Nevermind, I see what's going on now. This can probably be cleaned up, but I'm not going to pull that string right now. I've put this back.

ephraimbuddy replied:
Should we import the TaskNote and DagrunNote tables in this place instead?

jedcunningham replied:
Okay, I've pulled this out into #27869, just to keep things a little cleaner.

jedcunningham commented on airflow/models/dagrun.py:
```diff
@@ -85,6 +87,17 @@ class TISchedulingDecision(NamedTuple):
 
+def _creator_note(val):
+    """Custom creator for the ``note`` association proxy."""
+    # breakpoint()
```
```suggestion
```

jedcunningham commented on airflow/models/taskinstance.py:
```diff
@@ -313,6 +313,17 @@ def key(self) -> TaskInstanceKey:
 
+def _creator_note(val):
+    """Custom creator for the ``note`` association proxy."""
+    # breakpoint()
```
```suggestion
```

jedcunningham commented on airflow/www/views.py:
```diff
@@ -263,19 +263,20 @@ def dag_to_grid(dag, dag_runs, session):
-            TaskInstance.notes,
-            sqla.func.count(sqla.func.coalesce(TaskInstance.state, sqla.literal("no_status"))).label(
+            func.min(TaskInstanceNote.content).label("notes"),
+            func.count(sqla.func.coalesce(TaskInstance.state, sqla.literal("no_status"))).label(
```
```suggestion
            func.count(func.coalesce(TaskInstance.state, sqla.literal("no_status"))).label(
```

jedcunningham committed:
Apply suggestions from code review

jedcunningham committed:
Test that a user_id is set when creating note via api

jedcunningham committed:
Merge branch 'main' into notes-table

jedcunningham committed:
Fix static checks

jedcunningham committed:
Merge branch 'main' into notes-table

ephraimbuddy commented on airflow/models/taskinstance.py:
```diff
@@ -2676,6 +2684,52 @@ def from_dict(cls, obj_dict: dict) -> SimpleTaskInstance:
+    task_instance = relationship("TaskInstance", back_populates="task_instance_note")
+
+    __table_args__ = (
+        ForeignKeyConstraint(
```
We need to add explicit PrimaryKeyConstraint with a name here otherwise it will be different from what's on the migration file and will give issues for ORM created db if we want to migrate it
